### PR TITLE
feat(s3): dom based xml parser

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

This change proposes an XML document parser for S3 client based on DOMParser. So the browser clients do not need a `fast-xml-parser` dependency.

On runtimes where `DOMParser` is not available(RN, Node.js), it uses the same parser(based on `fast-xml-parser`) as AWS SDK client.

This change makes sure both XML parser populates the same results given XML protocol test cases.

#### Description of how you validated changes
Unit test

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/AwsClients/S3/utils/xmlParser.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { XMLParser } from 'fast-xml-parser';
```
Comment (cshfang) [inline packages/storage/src/AwsClients/S3/utils/xmlParser.ts:4]: Should this be an explicit dependency now?

File: packages/storage/src/AwsClients/S3/utils/xmlParser.browser.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	!hasNonNamespaceAttributes(node) &&
+	node.children.length === 0 &&
+	node.firstChild?.nodeType === Node.TEXT_NODE;
+
+const hasNonNamespaceAttributes = (node: Element): boolean => {
+	let hasAttributes = false;
+	for (let i = 0; i < node.attributes.length; i++) {
+		const attr = node.attributes[i];
+		if (!isNamespaceAttributeName(attr.nodeName)) {
+			hasAttributes = true;
```
Comment (cshfang) [inline packages/storage/src/AwsClients/S3/utils/xmlParser.browser.ts:86]: Nit: Can we simplify the control flow and just return true here? Then I think the `hasAttributes` and the `break` would be unneeded as well.

File: packages/storage/src/AwsClients/S3/utils/xmlParser.browser.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+const isElementNode = (node: Node): node is Element =>
+	node.nodeType === Node.ELEMENT_NODE;
+
+const isDocumentNode = (node: Node): node is Document =>
+	node.nodeType === Node.DOCUMENT_NODE;
+
+const isTextOnlyElementNode = (node: Element): boolean =>
+	!hasNonNamespaceAttributes(node) &&
```
Comment (cshfang) [inline packages/storage/src/AwsClients/S3/utils/xmlParser.browser.ts:77]: Given that this is the only application of the `hasNonNamespaceAttributes` function, I think the double negative here (`!` and `..Non..` in function name) is confusing. Could we update the function name to something like `hasOnlyNamespaceAttributes` and flip the logic? E.g.

```ts
const isTextOnlyElementNode = (node: Element): boolean =>
	hasOnlyNamespaceAttributes(node) &&
	node.children.length === 0 &&
	node.firstChild?.nodeType === Node.TEXT_NODE;

const hasOnlyNamespaceAttributes = (node: Element): boolean => {
	for (let i = 0; i < node.attributes.length; i++) {
		const { nodeName } = node.attributes[i];
		if (!isNamespaceAttributeName(nodeName)) {
			return false;
		}
	}
	return true;
};
```

**Review by elorzafe — COMMENTED** [review]
Looks good!

Thanks @AllanZhengYP 

Can we add this to the list of platform specific functionality, for V6 we want this to be part of amplify/react-native package

File: packages/storage/src/AwsClients/S3/utils/xmlParser.browser.ts [hunk 4]
```diff
@@ -74,20 +74,18 @@ const isDocumentNode = (node: Node): node is Document =>
 	node.nodeType === Node.DOCUMENT_NODE;
 
 const isTextOnlyElementNode = (node: Element): boolean =>
-	!hasNonNamespaceAttributes(node) &&
+	hasOnlyNamespaceAttribute(node) &&
```
Comment (cshfang) [inline packages/storage/src/AwsClients/S3/utils/xmlParser.browser.ts:77]: Nit
```suggestion
	hasOnlyNamespaceAttributes(node) &&
```

File: packages/storage/src/AwsClients/S3/utils/xmlParser.browser.ts [hunk 5]
```diff
@@ -74,20 +74,18 @@ const isDocumentNode = (node: Node): node is Document =>
...
 	node.children.length === 0 &&
 	node.firstChild?.nodeType === Node.TEXT_NODE;
 
-const hasNonNamespaceAttributes = (node: Element): boolean => {
-	let hasAttributes = false;
+const hasOnlyNamespaceAttribute = (node: Element): boolean => {
```
Comment (cshfang) [inline packages/storage/src/AwsClients/S3/utils/xmlParser.browser.ts:81]: Nit
```suggestion
const hasOnlyNamespaceAttributes = (node: Element): boolean => {
```

**Review by cshfang — APPROVED** [review]

**Pull request merged**